### PR TITLE
fix(api-key): prevent continuous traffic from extending rate-limit windows

### PR DESCRIPTION
**Pull request by bytaesu**
> [!IMPORTANT]
> This adds the nullable `rateLimitResetAt` field. It requires a database schema migration but does not introduce a breaking API change.

Replaces https://github.com/better-auth/better-auth/pull/8803
Closes https://github.com/better-auth/better-auth/issues/6035

---
## Summary by cubic
Fixes API key rate limiting by switching to first-request-anchored fixed windows in `@better-auth/api-key`. Continuous traffic no longer delays resets, and a nullable rateLimitResetAt tracks the window deadline.

- **Bug Fixes**
  - Use fixed windows; the first accepted request opens a window and sets the reset deadline.
  - Denials report tryAgainIn from the stable deadline; lastRequest no longer controls resets.
  - Concurrency-safe guards ensure only one window opens and increments don’t exceed the max.
  - Changing rate-limit config (enabled/max/window) clears the current window and count.
  - Added rateLimitResetAt across adapter, schema, API responses, and docs.
  - Preserve legacy counters by initializing resetAt from lastRequest without resetting count; after expiry, open a fresh fixed window.

- **Migration**
  - Add a nullable rateLimitResetAt column to the API keys table.
  - No breaking API changes; run your schema migration.

<sup>Written for commit ef3d1c638c59d4ef2e1b3efe298b597aa9ad68a4. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/better-auth/better-auth/pull/10747?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Review by ping-maxwell — APPROVED** [review]
LGTM, mostly looks the same as the original PR.

Just needs to check AI review comments.

File: packages/api-key/src/rate-limit.ts [hunk 1]
```diff
@@ -66,32 +75,74 @@ export function evaluateRateLimit(
...
+		const observedLastRequest = apiKey.lastRequest
+			? new Date(apiKey.lastRequest)
+			: null;
+
+		if (apiKey.requestCount > 0 && observedLastRequest) {
+			const resetAt = new Date(
+				observedLastRequest.getTime() + rateLimitTimeWindow,
+			);
+
+			if (now.getTime() < resetAt.getTime()) {
```
[the remark this answers is not in the document]
Comment (bytaesu) [reply packages/api-key/src/rate-limit.ts:96]: Fixed windows use `[start, resetAt)`, so `resetAt` begins the next window. Preserving the legacy `>` comparison would reject with `tryAgainIn: 0`. Both boundary cases are covered by tests.

File: packages/api-key/src/routes/update-api-key.ts [hunk 2]
```diff
@@ -427,6 +434,14 @@ export function updateApiKey({
 			if (rateLimitMax !== undefined) {
 				newValues.rateLimitMax = rateLimitMax;
 			}
+			if (
+				rateLimitEnabled !== undefined ||
+				rateLimitTimeWindow !== undefined ||
+				rateLimitMax !== undefined
+			) {
+				newValues.requestCount = 0;
```
[the remark this answers is not in the document]
Comment (bytaesu) [reply packages/api-key/src/routes/update-api-key.ts:442]: The minor bump is intentional. Although the field is nullable and API-compatible, database-backed users must apply a schema migration. Per `CONTRIBUTING.md`, patch bumps are for changes users do not need to know about. This change requires user action.

[the remark this answers is not in the document]
Comment (bytaesu) [reply packages/api-key/src/routes/update-api-key.ts:442]: @greptileai you also learn this https://github.com/better-auth/better-auth/pull/10747#discussion_r3752258474

**Comment by bytaesu** [comment]
Target v1.8
